phfs.c

```diff
@@ -13,61 +13,383 @@
  * %LICENSE%
  */
 
-#include "errors.h"
+#include "phfs.h"
+#include "devs.h"
 #include "hal.h"
+
+#include "phoenixd.h"
 #include "plostd.h"
-#include "phfs.h"
+#include "errors.h"
+
+
+#define SIZE_PHFS_HANDLERS    8
+#define SIZE_PHFS_FILES       10
+
+
 
-/* Configuration file for specific platform */
-#include "config.h"
+enum { phfs_prot_raw = 0, phfs_prot_phoenixd };
 
 
-phfs_handler_t phfs_handlers[PDN_NB];
+typedef struct {
+	char alias[8];
+	unsigned int major;
+	unsigned int minor;
+	unsigned int prot;
 
+	dev_handler_t *h;
+} phfs_handler_t;
 
-handle_t phfs_open(u16 pdn, const char *name, u32 flags)
+
+typedef struct {
+	char alias[32];
+	addr_t addr;
+	size_t size;
+} phfs_file_t;
+
+
+struct {
+	phfs_handler_t handlers[SIZE_PHFS_HANDLERS];
+	unsigned int hCnt;
+
+	phfs_file_t files[SIZE_PHFS_FILES];
+	unsigned int fCnt;
+} phfs_common;
+
+
+static const char *phfs_getProtName(unsigned int id)
 {
-	handle_t handle;
+	switch (id) {
+	case phfs_prot_raw:
+		return "raw";
 
-	if (pdn >= PDN_NB) {
-		handle.h = ERR_ARG;
-		return handle;
+	case phfs_prot_phoenixd:
+		return "phoenixd";
+
+	default:
+		return "";
 	}
+}
 
-	return phfs_handlers[pdn].open(phfs_handlers[pdn].dn, name, flags);
+
+static int phfs_setProt(phfs_handler_t *ph, const char *prot)
+{
+	if (prot == NULL || (plostd_strcmp(prot, "raw") == 0))
+		ph->prot = phfs_prot_raw;
+	else if (plostd_strcmp(prot, "phoenixd") == 0)
+		ph->prot = phfs_prot_phoenixd;
+	else
+		return ERR_ARG;
+
+	return ERR_NONE;
+}
+
+
+static int phfs_getHandlerId(const char *alias)
+{
+	int i;
+	phfs_handler_t *ph;
+
+	for (i = 0; i < phfs_common.hCnt; ++i) {
+		ph = &phfs_common.handlers[i];
+		if ((plostd_strcmp(alias, ph->alias) == 0))
+			return i;
+	}
+
+	return ERR_ARG;
 }
 
 
-s32 phfs_read(u16 pdn, handle_t handle, addr_t *pos, u8 *buff, u32 len)
+int phfs_regDev(const char *alias, unsigned int major, unsigned int minor, const char *prot)
 {
-	if (pdn >= PDN_NB)
+	size_t sz;
+	dev_handler_t *dh;
+	phfs_handler_t *ph;
+
+	if (alias == NULL)
+		return ERR_ARG;
+
+	if (phfs_common.hCnt >= SIZE_PHFS_HANDLERS) {
+		plostd_printf(ATTR_ERROR, "\nToo many devices");
+		return ERR_ARG;
+	}
+
+	if (devs_getHandler(major, minor, &dh) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%d.%d - device does not exist", major, minor);
 		return ERR_ARG;
+	}
+
+	/* Check if alias is already in use */
+	if (phfs_getHandlerId(alias) >= 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - alias is already in use", alias);
+		return ERR_ARG;
+	}
+
+	ph = &phfs_common.handlers[phfs_common.hCnt];
+	if (phfs_setProt(ph, prot) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - wrong protocol name\n\t use: \"%s\", \"%s\"", prot,
+		              phfs_getProtName(phfs_prot_raw), phfs_getProtName(phfs_prot_phoenixd));
+		return ERR_ARG;
+	}
+
+	sz = plostd_strlen(alias);
+	sz = (sz < (sizeof(ph->alias) - 1)) ? (sz + 1) : (sizeof(ph->alias) - 1);
+	hal_memcpy(ph->alias, alias, sz);
+	ph->alias[sz] = '\0';
```

Comment:
Currently, we don't have implementation of `strncpy`. We are going to  make modification in `plostd` module so it will be also a good time to add some extra functions.

Reply:
We have no implementation of strncpy and we don't plan to link libphoenix or internal compiler primitives, so I think this approach is correct.